x-pack/plugins/apm/public/components/app/correlations/use_failed_transactions_correlations.ts

```diff
@@ -0,0 +1,252 @@
+import { chunk, debounce } from 'lodash';
+
+import { IHttpFetchError, ResponseErrorBody } from 'src/core/public';
+
+import { EVENT_OUTCOME } from '../../../../common/elasticsearch_fieldnames';
+import { EventOutcome } from '../../../../common/event_outcome';
+import {
+  DEBOUNCE_INTERVAL,
+  DEFAULT_PERCENTILE_THRESHOLD,
+} from '../../../../common/correlations/constants';
+import type {
+  FailedTransactionsCorrelation,
+  FailedTransactionsCorrelationsResponse,
+} from '../../../../common/correlations/failed_transactions_correlations/types';
+
+import { callApmApi } from '../../../services/rest/createCallApmApi';
+
+import {
+  getInitialResponse,
+  getFailedTransactionsCorrelationsSortedByScore,
+  getReducer,
+  CorrelationsProgress,
+} from './utils/analysis_hook_utils';
+import { useFetchParams } from './use_fetch_params';
+
+// Overall progress is a float from 0 to 1.
+const LOADED_OVERALL_HISTOGRAM = 0.05;
+const LOADED_FIELD_CANDIDATES = LOADED_OVERALL_HISTOGRAM + 0.05;
+const LOADED_DONE = 1;
+const PROGRESS_STEP_P_VALUES = 0.9;
+
+export function useFailedTransactionsCorrelations() {
+  const abortCtrl = useRef(new AbortController());
+
+  const fetchParams = useFetchParams();
+
+  // This use of useReducer (the dispatch function won't get reinstantiated
+  // on every update) and debounce avoids flooding consuming components with updates.
+  const [response, setResponseUnDebounced] = useReducer(
+    getReducer<FailedTransactionsCorrelationsResponse & CorrelationsProgress>(),
+    getInitialResponse()
+  );
+  const setResponse = useMemo(
+    () => debounce(setResponseUnDebounced, DEBOUNCE_INTERVAL),
+    []
+  );
+
+  // We're using a ref here because otherwise the startFetch function might have
+  // a stale value for checking if the task has been cancelled.
+  const isCancelledRef = useRef(false);
+
+  const startFetch = useCallback(async () => {
+    abortCtrl.current.abort();
+    abortCtrl.current = new AbortController();
+    isCancelledRef.current = false;
+
+    setResponse({
+      ...getInitialResponse(),
+      isRunning: true,
+      // explicitly set these to undefined to override a possible previous state.
+      error: undefined,
+      failedTransactionsCorrelations: undefined,
+      percentileThresholdValue: undefined,
+      overallHistogram: undefined,
+      errorHistogram: undefined,
+      fieldStats: undefined,
+    });
+    setResponse.flush();
+
+    try {
+      // `responseUpdate` will be enriched with additional data with subsequent
+      // calls to the overall histogram, field candidates, field value pairs, correlation results
+      // and histogram data for statistically significant results.
+      const responseUpdate: FailedTransactionsCorrelationsResponse = {
+        ccsWarning: false,
+      };
+
+      // Initial call to fetch the overall distribution for the log-log plot.
+      const { overallHistogram, percentileThresholdValue } = await callApmApi({
```

Comment:
why are these requests all blocking? it looks like they could be executed in parallel?

Reply:
You're right, these don't depend on each other. I combined the first two with `Promise.all`. I left the third one that fetches `fieldCandidates` (then part of the analysis done for the table) to be called after that so we show the chart as soon as it's available. 97890cf